Expand ArmorStand API

willies952002 opened the pull request:
Add the following:
- Add proper methods for getting and setting items in both hands. Deprecates old methods
- Enable/Disable slot interactions

kashike commented on Spigot-API-Patches/0127-Expand-ArmorStand-API.patch:
```diff
@@ -0,0 +1,109 @@
++     * @see ArmorStand#getItemInMainHand()
++     // Paper end
+      */
+     ItemStack getItemInHand();
```
Please add the `@Deprecated` annotation.

kashike commented on Spigot-API-Patches/0127-Expand-ArmorStand-API.patch:
```diff
@@ -0,0 +1,109 @@
++     * @see ArmorStand#setItemInMainHand(ItemStack)
++     // Paper end
+      */
+     void setItemInHand(ItemStack item);
```
Please add the `@Deprecated` annotation.

kashike commented on Spigot-API-Patches/0127-Expand-ArmorStand-API.patch:
```diff
@@ -0,0 +1,109 @@
++     *
++     * @return the held item
++     */
++    ItemStack getItemInMainHand();
```
I'd like to see a method like https://github.com/PaperMC/Paper/blob/master/Spigot-Server-Patches/0091-Access-items-by-EquipmentSlot.patch#L18-L24:
```java
ItemStack getItemInHand(final org.bukkit.inventory.EquipmentSlot slot);
```

aikar replied:
I don't think we need these API's anymore with .getItem and setItem, it introduces confusion on "are these different", so let's just remove the hand specific ones.

Update the deprecated tags above too.

After that, I think we're good to merge.

kashike commented on Spigot-API-Patches/0127-Expand-ArmorStand-API.patch:
```diff
@@ -0,0 +1,109 @@
++     *
++     * @param item the item to hold
++     */
++    void setItemInMainHand(ItemStack item);
```
I'd like to see a method like https://github.com/PaperMC/Paper/blob/master/Spigot-Server-Patches/0091-Access-items-by-EquipmentSlot.patch#L39-L47:
```java
ItemStack setItemInHand(final org.bukkit.inventory.EquipmentSlot slot, final ItemStack item);
```

kashike commented on Spigot-API-Patches/0127-Expand-ArmorStand-API.patch:
```diff
@@ -0,0 +1,109 @@
++     *
++     * @param slots var-arg array of slots to lock
++     */
++    void setDisabledSlots(EquipmentSlot... slots);
```
```java
void addDisabledSlot(final EquipmentSlot... slots);

void removedDisabledSlot(final EquipmentSlot... slots);
```

kashike commented on Spigot-API-Patches/0127-Expand-ArmorStand-API.patch:
```diff
@@ -0,0 +1,146 @@
++     *
++     * @return integer
++     */
++    List<EquipmentSlot> getDisabledSlots();
```
Should be a `Set`.

kashike commented on Spigot-API-Patches/0127-Expand-ArmorStand-API.patch:
```diff
@@ -0,0 +1,146 @@
+@@ -1,8 +1,11 @@
+ package org.bukkit.entity;
+ 
++import org.bukkit.inventory.EquipmentSlot; // Paper
```
Avoid the import statement - fully qualify below.

kashike commented on Spigot-API-Patches/0127-Expand-ArmorStand-API.patch:
```diff
@@ -0,0 +1,146 @@
+ import org.bukkit.inventory.ItemStack;
+ import org.bukkit.util.EulerAngle;
+ 
++import java.util.Set; // Paper
```
Avoid the import statement - fully qualify below.

kashike commented on Spigot-Server-Patches/0315-Implement-Expanded-ArmorStand-API.patch:
```diff
@@ -0,0 +1,144 @@
++    @Override
++    public void removeDisabledSlots(org.bukkit.inventory.EquipmentSlot... slots) {
++        java.util.Set<org.bukkit.inventory.EquipmentSlot> disabled = getDisabledSlots();
++        disabled.removeAll(java.util.Arrays.asList(slots));
```
Just do a normal for loop here:
```java
for (final org.bukkit.inventory.EquipmentSlot slot : slots) disabled.remove(slot);
```

kashike commented on Spigot-Server-Patches/0315-Implement-Expanded-ArmorStand-API.patch:
```diff
@@ -0,0 +1,144 @@
++    @Override
++    public void addDisabledSlots(org.bukkit.inventory.EquipmentSlot... slots) {
++        java.util.Set<org.bukkit.inventory.EquipmentSlot> disabled = getDisabledSlots();
++        disabled.addAll(java.util.Arrays.asList(slots));
```
`java.util.Collections.addAll(disabled, slots);`

aikar commented on Spigot-Server-Patches/0315-Implement-Expanded-ArmorStand-API.patch:
```diff
@@ -0,0 +1,144 @@
++    }
++
++    @Override
++    public ItemStack getItemInHand(org.bukkit.inventory.EquipmentSlot slot) {
```
Actually, we should very well make this setItem/getItem(slot) and make all the slots work.

aikar commented on Spigot-API-Patches/0127-Expand-ArmorStand-API.patch:
```diff
@@ -0,0 +1,134 @@
++     */
++    void setItem(final org.bukkit.inventory.EquipmentSlot slot, final ItemStack item);
++
++    /**
```
I think all of the disabled slots API's should describe just what this API does so people don't have to go dig into technical wiki articles on what disabled slots do.

I don't even know!

willies952002 replied:
`This makes it so a player is unable to place, remove, or replace an item in the given slot(s)`

does this seem like it's not enough of a description?
For some reason I feel this needs a slightly longer explanation

aikar replied:
maybe something like "This makes it so a player is unable to interact with the Armor Stand to place, remove, or replace an item in the given slot(s)."

I assume the items fall if the armor stand is broken still, so may be worth noting that expectation, that once a slot is disabled, item retrieval can only be done on break?

willies952002 committed:
Expand ArmorStand API

Add the following:
- Add proper methods for getting and setting items. Deprecates old methods
- Enable/Disable slot interactions
